Applications/Tizen_native/CustomShortcut/src/data.c

```diff
@@ -43,6 +43,199 @@ int parse_route(const char *source, char **route, char **data) {
   return APP_ERROR_NONE;
 }
 
+void data_get_resource_path(const char *file, char *full_path, bool shared) {
+  char *root_path;
+  if (shared) {
+    root_path = app_get_shared_resource_path();
+  } else {
+    root_path = app_get_resource_path();
+  }
+
+  if (root_path == NULL) {
+    LOG_E("failed to get resource path");
+    return;
+  }
+
+  if (full_path == NULL) {
+    LOG_E("full_path is null");
+  }
+
+  snprintf(full_path, PATH_MAX, "%s%s", root_path, file);
+  LOG_D("resource path: %s", full_path);
+  free(root_path);
+}
+
+void data_get_data_path(const char *file, char *full_path) {
+  char *root_path;
+
+  root_path = app_get_data_path();
+
+  if (root_path == NULL) {
+    LOG_E("failed to get data path");
+    return;
+  }
+
+  if (full_path == NULL) {
+    LOG_E("full_path is null");
+    return;
+  }
+
+  snprintf(full_path, PATH_MAX, "%s%s", root_path, file);
+  LOG_D("data path: %s", full_path);
+  free(root_path);
+}
+
+static void _on_data_receive(ml_tensors_data_h data,
+                             const ml_tensors_info_h info, void *user_data) {
+  appdata_s *ad = (appdata_s *)user_data;
+
+  void *raw_data;
+  size_t data_size;
+  FILE *file;
+
+  pthread_mutex_lock(&ad->pipe_lock);
+
+  /// @note data_size written here will be overriden and it is intended.
+  int status = ml_tensors_data_get_tensor_data(data, 0, &raw_data, &data_size);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("get tensor data failed %d", status);
+    status = ml_tensors_info_get_tensor_size(info, -1, &data_size);
+    goto CLEAN;
+  }
+
+  file = fopen(ad->pipe_dst, ad->tries == 1 ? "wb" : "wb+");
+
+  if (file == NULL) {
+    LOG_E("cannot open file");
+  }
+
+  if (fwrite(raw_data, data_size, 1, file) < 0) {
```

Comment:
this will do fwrite when fopen failed?

Reply:
Add `goto CLEAN` when `file == NULL`

Reply:
What if return value of fwrite is positive, however less than data_size?
is that success case?

Reply:
It is error. Added the case :)